testsuite/tests/quotation/build_quotation.ml

```diff
@@ -0,0 +1,767 @@
+|}];;
+
+<[ function | Some x -> x | None -> 0 ]>;;
+[%%expect {|
+- : <[ int option -> int ]> expr = <[ function | Some (x) -> x | None -> 0 ]>
+|}];;
+
+<[ function | [] -> false | x::xs -> true ]>;;
+[%%expect {|
+- : <[ $ ('a) list -> bool ]> expr =
+<[ function | [] -> false | (::) (x, xs) -> true ]>
+|}];;
+
+<[ fun x d -> match x with | Some y -> y | None -> d ]>;;
+[%%expect {|
+- : <[ $ ('a) option -> $ ('a) -> $ ('a) ]> expr =
+<[ fun x d -> match x with | Some (y) -> y | None -> d ]>
+|}];;
+
+<[ fun l -> List.map (fun x -> 2 * x) l ]>;;
+[%%expect {|
+- : <[ int list -> int list ]> expr =
+<[ fun l -> Stdlib.List.map (fun x -> 2 * x) l ]>
+|}];;
+
+<[ fun (type a) (f : a -> a) (x : a) -> f (f x) ]>;;
+[%%expect {|
+- : <[ ($ ('a) -> $ ('a)) -> $ ('a) -> $ ('a) ]> expr =
+<[ fun (type a) (f : a -> a) (x : a) -> f (f x) ]>
+|}];;
+
+<[ fun x (type a) (f : a -> a * a) (g : int -> a) -> f (g x) ]>;;
+[%%expect {|
+- : <[
+     int -> ($ ('a) -> $ ('a) * $ ('a)) -> (int -> $ ('a)) -> $ ('a) * $ ('a)
+     ]>
+    expr
+= <[ fun x (type a) (f : a -> (a) * (a)) (g : int -> a) -> f (g x) ]>
+|}];;
+
+<[ fun (f : 'a. 'a -> 'a) -> f f ]>;;
+[%%expect {|
+- : <[ ('a. 'a -> 'a) -> $ ('b) -> $ ('b) ]> expr =
+<[ fun (f : 'a. 'a -> 'a) -> f f ]>
+|}];;
+
+<[ fun x -> fun x -> fun x -> 42 ]>;;
+[%%expect {|
+- : <[ $ ('a) -> $ ('b) -> $ ('c) -> int ]> expr =
+<[ fun x -> fun x__1 -> fun x__2 -> 42 ]>
+|}];;
+
+<[ fun x -> fun x -> fun x__1 -> 42 ]>;;
+[%%expect {|
+- : <[ $ ('a) -> $ ('b) -> $ ('c) -> int ]> expr =
+<[ fun x -> fun x__1 -> fun x__2 -> 42 ]>
+|}];;
+
+<[ let z = 10 in z ]>;;
+[%%expect {|
+- : <[ int ]> expr = <[ let z = 10 in z ]>
+|}];;
+
+<[ let (x, y) = (42, 100) in x + y ]>;;
+[%%expect {|
+- : <[ int ]> expr = <[ let (x, y) = (42, 100) in x + y ]>
+|}];;
+
+<[ let Some x = Some "foo" in x ]>;;
+[%%expect {|
+Line 1, characters 3-31:
+1 | <[ let Some x = Some "foo" in x ]>;;
+       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Warning 8 [partial-match]: this pattern-matching is not exhaustive.
+Here is an example of a case that is not matched:
+None
+
+- : <[ string ]> expr = <[ match Some "foo" with | Some (x) -> x ]>
+|}];;
+
+<[ let x::xs = [1; 2; 3] in x ]>;;
+[%%expect {|
+Line 1, characters 3-29:
+1 | <[ let x::xs = [1; 2; 3] in x ]>;;
+       ^^^^^^^^^^^^^^^^^^^^^^^^^^
+Warning 8 [partial-match]: this pattern-matching is not exhaustive.
+Here is an example of a case that is not matched:
+[]
+
+- : <[ int ]> expr =
+<[ match (::) (1, ((::) (2, ((::) (3, []))))) with | (::) (x, xs) -> x ]>
+|}];;
+
+<[ let x::xs = [1; 2; 3] in xs ]>;;
+[%%expect {|
+Line 1, characters 3-30:
+1 | <[ let x::xs = [1; 2; 3] in xs ]>;;
+       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Warning 8 [partial-match]: this pattern-matching is not exhaustive.
+Here is an example of a case that is not matched:
+[]
+
+- : <[ int list ]> expr =
+<[ match (::) (1, ((::) (2, ((::) (3, []))))) with | (::) (x, xs) -> xs ]>
```

Comment:
We shouldn't put too much effort into nice printing, but it is probably worth including a case to print list literals using the nice syntax.